### ISSUE TITLE
LightGBMRanker stopped training (No further splits with positive gain)

### ISSUE THREAD

**Issue by ElysiumFan086** [Azure/mmlspark#867]
I have encountered some trouble when training Lambda rank model in Spark with LightGBMRanker in mmlspark.

With the same training data, training results in Spark and local machine are different:

1. In Spark with mmlspark, training log is shown as below:

   ```
   [LightGBM] [Info] Total Bins 15570
   [LightGBM] [Info] Number of data: 6931159, number of used features: 101
   [LightGBM] [Warning] No further splits with positive gain, best gain: -inf
   [LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements
   ```

2. In my local machine, where LightGBM is installed with conda, training work finished successfully.

I am not sure what makes above difference. I have tried to modify some parameters like `maxPosition`  and partition number in spark, which had proved to be of no use. And also setting `min_data_in_leaf`  smaller was also proposed, but I didn't find any access to set this parameter in mmlspark interface of LightGBMRanker.

If anyone had experience of solving similar problem, it will be very appreciate of you to give me some advice.

Following lists some parameters used to initialize the ranker:
```
        model = LightGBMRanker(
            parallelism='data_parallel',
            objective='lambdarank',
            boostingType='gbdt',
            numIterations=500,
            learningRate=0.1,
            numLeaves=1023,
            maxDepth=10,
            earlyStoppingRound=0,
            maxPosition=8,
            minSumHessianInLeaf=0.001,
            lambdaL1=0.01,
            lambdaL2=0.01,
            isProvideTrainingMetric=True,
            defaultListenPort=49650,
            featuresCol='features',
            groupCol='query_id',
            labelCol='label',
            numBatches=0,
            timeout=600000.0,
            verbosity=1).fit(df)
```

**Comment by imatiach-msft** [comment]
hi @ElysiumFan086 I wonder if this is similar to this issue (based on a quick search):
https://github.com/microsoft/LightGBM/issues/2953
you can also try to post this to the lightgbm forum.
Is the dataset private?  Can you create a small repro with either a toy dataset or the one you are using (if it is not private)?  That may be the best way to debug this.

**Comment by imatiach-msft** [comment]
see some other similar issues:
https://github.com/microsoft/LightGBM/issues/2597
another possibly related post:
https://github.com/microsoft/LightGBM/issues/2239

Have you tried the latest code from master?  I just updated to the latest LightGBM on master recently.  This issue may have already been fixed.

**Comment by ElysiumFan086** [comment]
Thank you for your advice, and due to privacy policy it is inconvenient to open the dataset here. But I have check the dataset, and It seems that for items, in amount of queries, have same ranking labels, which may result from my labeling strategy, and I will check it carefully.
For example, ranking label in Query may be:
```
Q1: 4, 4, 4, 4, 4, 0, 0 
Q2: 2, 2, 2, 2, 2, 2, 4
Q3: 1, 1, 1, 1, 4, 2
```

But still I have several problems to trouble you:
1. For mmlspark, is there any way to modify `min_data_in_leaf` parameter for LightGBMRanker;
2. Although the labels of training data are a little unreasonable, yet it still cannot explain why it works on a single machine;
3. I am going to try the latest version, which asks me to upload the jar packages of mmlspark and related LigthGBM to our private platform. So is it convenient of you to give a URL to download the needed jars;

I hope these will not bother you much, and thank you again!

**Comment by imatiach-msft** [comment]
@ElysiumFan086 
1.) That parameter should already be available on latest master:
https://github.com/Azure/mmlspark/blob/master/src/main/scala/com/microsoft/ml/spark/lightgbm/LightGBMParams.scala#L357
2.) I'm guessing that either the native code might be different, or the distributed data partitioning code causes the logic to be slightly different to cause the issue - either way the fix will probably be in the LightGBM codebase
3.) I think this is latest version:
Maven Coordinates

com.microsoft.ml.spark:mmlspark_2.11:1.0.0-rc1-86-05c25aad-SNAPSHOT

Maven Resolver

https://mmlspark.azureedge.net/maven

This comes from this PR build:
https://github.com/Azure/mmlspark/pull/866/checks?check_run_id=715637908

**Comment by ElysiumFan086** [comment]
Thanks~~ @imatiach-msft I have tried the latest version, but still face the same problem. Currently, we have no idea about this issue, and what we can do is training on local machine, which just takes me more time.

**Comment by wil70** [comment]
Hello, Any update on this? TY

**Comment by ElysiumFan086** [comment]
This is RongFan. I have received you mail just now, and I will read it soon.Thank you!
